File: scripts/bindings/ggplot_default_style_palettes.py
"""Reference shape/linetype default selection through actual portable hosts."""
import json,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True);records=[]
cases=json.loads((ROOT/'fixtures/parity/ggplot2/default-style-palettes.json').read_text())['cases'];registry=c.ExtensionRegistry.example()
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
def theme(t):return c.theme().scale_palettes({f"palette.{t['channel']}.discrete":{'operation':{'id':'example.scale_palette','version':'1'},'parameters':{'mode':'constant','channel':'size' if t['channel']=='shape' else 'linetype'}}}) if t['theme_mode']=='supplied' else c.theme()
def layer(t):
 p=(c.points() if t['channel']=='shape' else c.rule()).name('marks')
 if t['route']!='automatic':
  scale={'function':{'Ordinal':{'domain':[],'range':[],'unknown':{'Explicit':None}}},'training':'Eligible','ggplot':{'Discrete':{'limits':None,'levels':None,'drop':True,'na_translate':True,'palette':{'Shape':{'solid':t['route']!='hollow'}} if t['channel']=='shape' else 'LineType'}}}
  if t['route']=='constructor':scale['palette_theme_aesthetics']=[t['channel']]
  p=p.value_scale('Shape' if t['channel']=='shape' else 'LineType','v',scale)
 return p
def data_for(t):return c.Data.columns({'x':c.column(list(map(float,range(len(t['inputs'])))),kind='float64'),'end':c.column([float(i)+.5 for i in range(len(t['inputs']))],kind='float64'),'v':c.column(t['inputs'],kind='string').nullable(True)},name='data')
def build(t,d):
 m=c.aes().x('x').y(1.);m=m.shape('v') if t['channel']=='shape' else m.x2('end').y2(1.).linetype('v')
 return c.plot(d).profile('Ggplot2_4_0_3').with_registry(registry).aes(m).layer(layer(t)).theme(theme(t)).build()
def check(t,chart):
 actual=chart.semantics()['layers'][0];rows=actual.get('aesthetics',[]);wanted=[v for v in t['result']['mapped'] if v is not None or t['channel']=='linetype'];key='Shape' if t['channel']=='shape' else 'LineType';assert len(rows)==len(wanted),(t,rows)
 for row,v in zip(rows,wanted):assert row[key]==({'kind':'Missing'} if v is None else {'kind':'Number' if key=='Shape' else 'Text','value':v}),(t,row,v)
 return {'aesthetics':rows,'styles':actual.get('styles',[])}
def matching(x,t):return all(x[k]==t[k] for k in ('channel','route','theme_mode'))
for index,t in enumerate(cases):
 owned=[]
 try:
  d=data_for(t);owned.append(d);p=build(t,d);owned.append(p);wire=p.to_json();restored=c.Plot.from_json(wire,registry);owned.append(restored);assert restored.to_json()==wire
  for state in ('original','layer_edit','theme_edit'):
   expected=t
   if state=='original':current=restored
   elif state=='layer_edit':current=restored.edit().layer('marks',layer(t)).build();owned.append(current)
   else:
    expected=next(x for x in cases if x['channel']==t['channel'] and x['route']==t['route'] and x['population']==t['population'] and x['theme_mode']!=t['theme_mode']);current=restored.edit().theme(theme(expected)).build();owned.append(current)
   chart=current.chart();owned.append(chart);records.append({'index':index,'state':state,**check(expected,chart)});assert restored.to_json()==wire
  if t['population']=='ordinary':
   request=output.request(restored,options);owned.append(request);frame=request.prepare();owned.append(frame);scene=frame.scene()
   for fmt in ('svg','pdf','png'):(out/f"{t['channel']}-{t['route']}-{t['theme_mode']}.{fmt}").write_bytes(frame.export(fmt))
   chart=restored.chart();owned.append(chart)
   for population in ('missing','empty','ordinary'):
    expected=next(x for x in cases if matching(x,t) and x['population']==population);replacement=data_for(expected);owned.append(replacement);tx=chart.transaction();owned.append(tx);builder=tx.replace(d,replacement);owned.append(builder);update=builder.build();owned.append(update);assert 'Applied' in chart.commit(update)
    batch_plot=build(expected,replacement);owned.append(batch_plot);batch=batch_plot.chart();owned.append(batch);actual=check(expected,chart);assert actual==check(expected,batch);assert restored.to_json()==wire and frame.scene()==scene;records.append({'index':index,'state':'replacement','population':population,**actual})
 finally:
  for obj in reversed(owned):obj.dispose()
assert len(records)==144,len(records)
(out/'records.json').write_text(json.dumps(records,sort_keys=True));options.dispose();output.dispose();registry.dispose();print('PASS 144 style palette states and 36 publication files')

File: crates/chart-core/tests/ggplot_default_style_palettes.rs
//! FIX-GG04: automatic shape/linetype fallback versus explicit shape palettes.
use chart_core::{
    ChartResult, Revision,
    grammar::*,
    interpolate::{Number, Value},
    prelude::*,
    scales::*,
};
use serde_json::{Value as Json, json};
use std::sync::Arc;
struct Constant;
impl CustomScalePalette for Constant {
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("test.style_palette", Revision::new(1), true)
    }
    fn validate(&self, _: &Json) -> ChartResult<()> {
        Ok(())
    }
    fn evaluate(&self, input: ScalePaletteInput<'_>) -> ChartResult<ScalePaletteOutput> {
        let ScalePaletteDomain::Count(n) = input.domain else {
            panic!("discrete palette")
        };
        let value = if input.parameters["channel"] == "shape" {
            Value::Number(Number(3.))
        } else {
            Value::Text("22".into())
        };
        Ok(ScalePaletteOutput {
            values: Some(vec![value; n]),
            names: None,
        })
    }
}
#[test]
fn default_style_palettes_match_36_reference_draws() {
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/default-style-palettes.json"
    ))
    .unwrap();
    let cases = fixture["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 36);
    for t in cases {
        assert!(t["result"].get("error").is_none(), "{t}");
        let shape = t["channel"] == "shape";
        let channel = t["channel"].as_str().unwrap();
        let values = t["inputs"].as_array().unwrap();
        let d = Data::columns()
            .column("x", (0..values.len()).map(|i| i as f64).collect::<Vec<_>>())
            .column(
                "end",
                (0..values.len())
                    .map(|i| i as f64 + 0.5)
                    .collect::<Vec<_>>(),
            )
            .column(
                "v",
                values
                    .iter()
                    .map(|v| v.as_str().map(str::to_owned))
                    .collect::<Vec<_>>(),
            )
            .build()
            .unwrap();
        let mut registry = ExtensionRegistry::new();
        registry.register_scale_palette(Arc::new(Constant)).unwrap();
        let mapping = if shape {
            aes().x("x").y(1.).shape("v")
        } else {
            aes().x("x").y(1.).x2("end").y2(1.).linetype("v")
        };
        let mut layer = if shape { points() } else { rule() };
        if t["route"] != "automatic" {
            let palette = if shape {
                GgplotDiscretePalette::Shape {
                    solid: t["route"] != "hollow",
                }
            } else {
                GgplotDiscretePalette::LineType
            };
            let mut scale =
                MappedScaleSpec::authored(ScaleFunctionSpec::Ordinal(OrdinalSpec::default()))
                    .with_ggplot(GgplotScalePolicy::Discrete {
                        empty_population: false,
                        limits: None,
                        levels: None,
                        drop: true,
                        na_translate: true,
                        palette,
                    })
                    .unwrap();
            if t["route"] == "constructor" {
                scale = scale.with_theme_palette(vec![channel.into()]).unwrap();
            }
            layer = layer.value_scale(
                if shape {
                    ValueAesthetic::Shape
                } else {
                    ValueAesthetic::LineType
                },
                "v",
                scale,
            );
        }
        let mut draft = plot(d)
            .profile(Profile::Ggplot2_4_0_3)
            .extensions(Arc::new(registry))
            .aes(mapping)
            .layer(layer);
        if t["theme_mode"] == "supplied" {
            draft = draft.theme(
                theme().scale_palettes(
                    [(
                        format!("palette.{channel}.discrete"),
                        ScalePaletteOperation {
                            operation: OperationRef {
                                id: "test.style_palette".into(),
                                version: Revision::new(1),
                            },
                            parameters: json!({"channel":channel}),
                        },
                    )]
                    .into(),
                ),
            );
        }
        let p = draft.build().unwrap();
        let prepared = p.chart().unwrap().prepare().unwrap();
        let marks = prepared.layers()[0].marks();
        let wanted = t["result"]["mapped"].as_array().unwrap();
        if shape {
            assert_eq!(
                marks.len(),
                wanted.iter().filter(|v| !v.is_null()).count(),
                "{t}"
            );
            for (m, v) in marks.iter().zip(wanted.iter().filter(|v| !v.is_null())) {
                assert_eq!(
                    m.aesthetics.get(&ValueAesthetic::Shape),
                    Some(&Value::Number(Number(v.as_f64().unwrap()))),
                    "{t}"
                );
            }
        } else {
            assert_eq!(marks.len(), wanted.len(), "{t}");
            for (m, v) in marks.iter().zip(wanted) {
                let expected = if let Some(v) = v.as_str() {
                    LineType::parse(v).unwrap()
                } else {
                    LineType::Blank
                };
                assert_eq!(m.style.line_type, Some(expected), "{t}");
            }
        }
    }
}
